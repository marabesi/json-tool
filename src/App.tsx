import { useState } from 'react'
import JsonEditor from './components/JsonEditor'
import Label from './components/Label';
import Formatter from './core/formatter'

function App() {
  const [originalJson, setOriginalResult] = useState('');
  const [result, setResult] = useState('');

  const onJsonChange = async (value: string) => {
    const format = new Formatter(value);
    const result = await format.format();

    setOriginalResult(value);
    setResult(result);
  }

  return (
    <div className="h-screen bg-gray-500 p-5">
      <div className="flex h-3/6">
        <div className="w-3/6 flex flex-col h-full m-1">
          <Label>place your json here</Label>
          <JsonEditor
            input={originalJson}
            onChange={eventValue => onJsonChange(eventValue.value)}
            data-testid="json"
            className="original"
          />
        </div>
        <div className="w-3/6 flex flex-col h-full m-1">
          <Label>result</Label>
          <JsonEditor
            input={result}
            className="result"
          />
        </div>
      </div>
    </div>
  );
}

export default App;
