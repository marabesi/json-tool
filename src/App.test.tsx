import { fireEvent, render, screen, act } from '@testing-library/react';
import App from './App';
import userEvent from '@testing-library/user-event';
import { Blob } from 'buffer';

describe('json utility', () => {

  test('renders place your json here label', () => {
    render(<App />);
    const placeJsonLabel = screen.getByTestId('label-json');
    expect(placeJsonLabel).toBeInTheDocument();
  });

  test('renders result label', () => {
    render(<App />);
    const resultLabel = screen.getByTestId('label-result');
    expect(resultLabel).toBeInTheDocument();
  });

  test('renders resulting formatted json label', () => {
    render(<App />);
    const resultLabel = screen.getByText(/result/i);
    expect(resultLabel).toBeInTheDocument();
  });

  test('error message is hidden by default', () => {
    render(<App />);
    const errorLabel = screen.queryByTestId(/error/);
    expect(errorLabel).toBeNull();
  });

  test.each([
    ['{}', '{}'],
    ['{"a": "b"}', '{"a": "b"}'],
  ])('place %s text in the editor and receive %s', async (input, expected) => {
    render(<App />);

    const editor = screen.getByTestId('json');

    await act(async () => {
      await fireEvent.change(editor, {target: { value: input }});
    });

    const result = screen.getByTestId('result');

    expect(result.nodeValue).toMatchSnapshot(expected);
  });

  test('inform error when json is invalid', async () => {
    render(<App />);

    const editor = screen.getByTestId('json');

    await act(async () => {
      fireEvent.change(editor, {target: { value: 'bla bla' }});
    });

    const result = screen.getByTestId('error');

    expect(result.innerHTML).toEqual('invalid json');
  });

  test.each([
    ['bla bla', '{}'],
    ['not a json', ''],
  ])('hides the error after a valid json is given', async (originalCode: string, afterChangeCode: string) => {
    render(<App />);

    const editor = screen.getByTestId('json');

    await act(async () => {
      fireEvent.change(editor, {target: { value: originalCode }});
    });

    await act(async () => {
      fireEvent.change(editor, {target: { value: afterChangeCode }});
    });

    const result = screen.queryByTestId('error');

    expect(result).toBeNull();
  });

  test('should paste json string from copy area into the editor', async () => {
    render(<App />);

    const editor = screen.getByTestId('json');

    await act(async () => {
      userEvent.paste(editor, '{}');
    });

    const result = screen.getByTestId('result');

    expect(result).toHaveValue('{}');
  });

  test('should paste json string from copy area into the editor on clicking the button', async () => {
    render(<App />);

    Object.assign(global.navigator,
      {
        clipboard :{
          async read() {
            const blob = new Blob([JSON.stringify({})], { type: 'text/plain' });

            return Promise.resolve([
              {
                [blob.type]: blob,
                types: [ blob.type ],
                getType: () => blob
              }
            ]);
          }
        }
      });

    await act(async () => {
      const fromClipboard = screen.getByTestId('paste-from-clipboard');
      fromClipboard.click();
    });

    expect(screen.getByTestId('json')).toHaveValue('{}');
    expect(screen.getByTestId('result')).toHaveValue('{}');
  });

  test('should clean editors once clean is clicked', async () => {
    render(<App />);

    const editor = screen.getByTestId('json');

    await act(async () => {
      userEvent.paste(editor, '{}');
    });

    await act(async () => {
      userEvent.click(screen.getByTestId('clean'));
    });

    const result = screen.getByTestId('result');

    expect(editor).toHaveValue('');
    expect(result).toHaveValue('');
  });
});
