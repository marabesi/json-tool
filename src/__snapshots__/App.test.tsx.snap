// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`json utility Editors place {"a": "b"} text in the editor and receive {"a": "b"}: {"a": "b"} 1`] = `null`;

exports[`json utility Editors place {} text in the editor and receive {}: {} 1`] = `null`;
